src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java

```diff
@@ -1,78 +1,298 @@
 package net.minecraftforge.event.entity.player;
 
-import static net.minecraftforge.fml.common.eventhandler.Event.Result.DEFAULT;
-import static net.minecraftforge.fml.common.eventhandler.Event.Result.DENY;
+import com.google.common.base.Preconditions;
+import net.minecraft.entity.Entity;
 import net.minecraft.entity.player.EntityPlayer;
-import net.minecraft.util.math.BlockPos;
+import net.minecraft.item.ItemStack;
 import net.minecraft.util.EnumFacing;
+import net.minecraft.util.EnumHand;
+import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.Vec3d;
 import net.minecraft.world.World;
+import net.minecraftforge.common.MinecraftForge;
 import net.minecraftforge.fml.common.eventhandler.Cancelable;
+import net.minecraftforge.fml.relauncher.Side;
+
+import static net.minecraftforge.fml.common.eventhandler.Event.Result.DEFAULT;
+import static net.minecraftforge.fml.common.eventhandler.Event.Result.DENY;
 
 /**
  * PlayerInteractEvent is fired when a player interacts in some way.
- * <br>
- * This event is fired whenever a player interacts in
- * Minecraft#func_147121_ag(),
- * NetHandlerPlayServer#processPlayerBlockPlacement(C08PacketPlayerBlockPlacement),
- * ItemInWorldManager#activateBlockOrUseItem(EntityPlayer, World, ItemStack, int, int, int, int, float, float, float),
- * ItemInWorldManager#onBlockClicked(int, int, int, int). <br>
- * <br>
- * This event is fired via the {@link ForgeEventFactory#onPlayerInteract(EntityPlayer, Action, BlockPos, EnumFacing)}.
- * <br>
- * {@link #action} contains the Action the player performed durin this interaction. <br>
- * {@link #pos} contains the coordinate of where this event occurred.<br>
- * {@link #face} contains the face of the block that was interacted with. May be null if unknown. <br>
- * {@link #world} contains the world in which this event is occurring. <br>
- * <br>
- * This event is {@link Cancelable}.<br>
- * If this event is canceled, the player does not interact.<br>
- * <br>
- * This event does not have a result. {@link HasResult}<br>
- * <br>
- * This event is fired on the {@link MinecraftForge#EVENT_BUS}.
+ * All subclasses are fired on {@link MinecraftForge#EVENT_BUS}.
+ * See the individual documentation on each subevent for more details.
  **/
-@Cancelable
 public class PlayerInteractEvent extends PlayerEvent
 {
-    public static enum Action
+    private final EnumHand hand;
+    private final ItemStack stack;
+    private final BlockPos pos;
+    private final EnumFacing face;
+
+    private PlayerInteractEvent(EntityPlayer player, EnumHand hand, ItemStack stack, BlockPos pos, EnumFacing face)
+    {
+        super(Preconditions.checkNotNull(player, "Null player in PlayerInteractEvent!"));
+        this.hand = Preconditions.checkNotNull(hand, "Null hand in PlayerInteractEvent!");
+        this.stack = stack;
+        this.pos = Preconditions.checkNotNull(pos, "Null position in PlayerInteractEvent!");
+        this.face = face;
```

Comment:
Is face nullable? If so, might be a good idea to mark the parameter, field, and accessor with `@Nullable`.